src/ctl/ctl.c

```diff
@@ -0,0 +1,579 @@
+	 * query has been handled.
+	 */
+    struct ctl_indexes indexes;
+    PMDK_SLIST_INIT(&indexes);
+
+    int ret = -1;
+
+    const struct ctl_node *n = ctl_find_node(CTL_NODE(global), name, &indexes);
+
+    if (n == NULL && ctl) {
+        ctl_delete_indexes(&indexes);
+        n = ctl_find_node(ctl->root, name, &indexes);
+    }
+
+    if (n == NULL || n->type != CTL_NODE_LEAF || n->cb[type] == NULL) {
+        printf("invalid query entry point %s\n", name);
+        errno = EINVAL;
+        goto out;
+    }
+
+    ret = ctl_exec_query[type](ctx, n, source, arg, &indexes);
+
+out:
+    ctl_delete_indexes(&indexes);
+
+    return ret;
+}
+
+/*
+ * ctl_register_module_node -- adds a new node to the CTL tree root.
+ */
+void ctl_register_module_node(struct ctl *c, const char *name,
+                              struct ctl_node *n) {
+    struct ctl_node *nnode = c == NULL
+                                 ? &CTL_NODE(global)[ctl_global_first_free++]
+                                 : &c->root[c->first_free++];
+
+    nnode->children = n;
+    nnode->type = CTL_NODE_NAMED;
+    nnode->name = name;
+}
+
+/*
+ * ctl_parse_query -- (internal) splits an entire query string
+ *	into name and value
+ */
+static int ctl_parse_query(char *qbuf, char **name, char **value) {
+    if (qbuf == NULL) {
+        return -1;
+    }
+
+    char *sptr;
+    *name = strtok_r(qbuf, CTL_NAME_VALUE_SEPARATOR, &sptr);
+    if (*name == NULL) {
+        return -1;
+    }
+
+    *value = strtok_r(NULL, CTL_NAME_VALUE_SEPARATOR, &sptr);
+    if (*value == NULL) {
+        return -1;
+    }
+
+    /* the value itself mustn't include CTL_NAME_VALUE_SEPARATOR */
+    char *extra = strtok_r(NULL, CTL_NAME_VALUE_SEPARATOR, &sptr);
+    if (extra != NULL) {
+        return -1;
+    }
+
+    return 0;
+}
+
+/*
+ * ctl_load_config -- executes the entire query collection from a provider
+ */
+static int ctl_load_config(struct ctl *ctl, void *ctx, char *buf) {
+    int r = 0;
+    char *sptr = NULL; /* for internal use of strtok */
+    char *name;
+    char *value;
+
+    // ASSERTne(buf, NULL);
+
+    char *qbuf = strtok_r(buf, CTL_STRING_QUERY_SEPARATOR, &sptr);
+    while (qbuf != NULL) {
+        r = ctl_parse_query(qbuf, &name, &value);
+        if (r != 0) {
+            printf("failed to parse query %s\n", qbuf);
+            return -1;
+        }
+
+        r = ctl_query(ctl, ctx, CTL_QUERY_CONFIG_INPUT, name, CTL_QUERY_WRITE,
+                      value);
+        printf("ctl_query: %s\n", name);
+
+        if (r < 0 && ctx != NULL) {
+            return -1;
+        }
+
+        qbuf = strtok_r(NULL, CTL_STRING_QUERY_SEPARATOR, &sptr);
+    }
+
+    return 0;
+}
+
+/*
+ * ctl_load_config_from_string -- loads obj configuration from string
+ */
+int ctl_load_config_from_string(struct ctl *ctl, void *ctx,
+                                const char *cfg_string) {
+    // LOG(3, "ctl %p ctx %p cfg_string \"%s\"", ctl, ctx, cfg_string);
```

Comment:
`LOG` could be replaced with our `LOG_*` macros

Reply:
Temporarily deleted

Reply:
I'll leave this issue open, for LOGs to come back, when ready

Reply:
bump, pls add LOGs

Reply:
It's not a scope of this PR.

Reply:
a few places with the old  `LOG(` were removed, just please bring them back using our logger